flax/struct.py

```diff
@@ -70,66 +70,75 @@ def __apply__(self, *args):
 
   Args:
     clz: the class that will be transformed by the decorator.
+    frozen: whether to freeze the dataclass (default=True). WARNING: mutating
+      instance of the dataclass from within a module is not allowed and can
```

Comment:
I'm not sure the warning should be specifically about modules? Seems like you might run into problems with mutation even outside a module if you were doing some other JAX operation. Conversely, doing this inside a module might be safe as long as you knew what you were doing (i.e. locally creating, then mutating, a `flax.struct.dataclass` object you define yourself).

It seems to me the issue is more about the scope of your mutations, in the sense that you shouldn't expect mutating a dataclass inside any (jax-traced) function to have any effect outside that function. Roughly I think the semantics are something like "assume any argument value in your function might be a (possibly abstract) copy of the actual input argument (so changes will be lost), and don't ever put local traced values into global objects, instead make sure to return the object with your modifications".

(I might be missing something regarding module vs not-module, though; someone please correct me if I'm wrong!)

Reply:
Ya you're right, I wrote the docstring thinking of a Flax neural net user but ofc its Jax that's the issue. I just wanted to include some kind of warning in the docstring, but happy to just change it to some generic warning about only setting if you are aware of the caveats of mutation.

Reply:
Updated.